src/main/java/com/mongodb/kafka/connect/source/schema/BsonDocumentToSchema.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.kafka.connect.source.schema;
+
+import static java.lang.String.format;
+
+import java.util.List;
+import java.util.Objects;
+
+import org.apache.kafka.connect.data.Decimal;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Timestamp;
+
+import org.bson.BsonValue;
+
+public final class BsonDocumentToSchema {
+
+  private static final Schema DEFAULT_INFER_SCHEMA_TYPE = Schema.STRING_SCHEMA;
+  public static final String SCHEMA_NAME_TEMPLATE = "inferred_name_%s";
+
+  public static Schema inferSchema(final BsonValue bsonValue) {
+    switch (bsonValue.getBsonType()) {
+      case BOOLEAN:
+        return Schema.BOOLEAN_SCHEMA;
+      case INT32:
+        return Schema.INT32_SCHEMA;
+      case INT64:
+        return Schema.INT64_SCHEMA;
+      case DOUBLE:
+        return Schema.FLOAT64_SCHEMA;
+      case DECIMAL128:
+        return Decimal.schema(bsonValue.asDecimal128().getValue().bigDecimalValue().scale());
+      case DATE_TIME:
+      case TIMESTAMP:
+        return Timestamp.SCHEMA;
+      case DOCUMENT:
+        SchemaBuilder builder = SchemaBuilder.struct();
+        bsonValue.asDocument().forEach((k, v) -> builder.field(k, inferSchema(v)));
+        builder.name(generateName(builder));
+        return builder.build();
+      case ARRAY:
+        List<BsonValue> values = bsonValue.asArray().getValues();
+        Schema firstItemSchema =
+            values.isEmpty() ? DEFAULT_INFER_SCHEMA_TYPE : inferSchema(values.get(0));
+        if (values.isEmpty()
+            || values.stream().anyMatch(bv -> !Objects.equals(inferSchema(bv), firstItemSchema))) {
+          return SchemaBuilder.array(DEFAULT_INFER_SCHEMA_TYPE).build();
```

Comment:
So if there is an array with mixed types, like `[1, "one", true]` then the schema type will be array of string?  Will that force all the values to be converted to a string as well?  Is that desirable?

Reply:
As there is no option to support multiple types in an array, it is the least undesirable solution.

Reply:
Does this mean we will convert that array to `["1", "one", "true"]`?

Reply:
Yes, String or Json representation of values is the lowest type

Reply:
Scary, but also probably uncommon